lib/ueberauth/strategy/slack/slack.ex

```diff
@@ -100,161 +127,6 @@ defmodule Ueberauth.Strategy.Slack do
 
   def challenge(_, _), do: {:error, :invalid_params}
 
-  @impl true
-  @spec authenticate(Ueberauth.Strategy.provider_name(), authenticate_params, options) ::
-          {:ok, Auth.t()} | {:error, Failure.t()}
-  def authenticate(provider, %{conn: conn, query: %{"code" => _code} = params}, opts) do
-    auth_url = request_uri(conn)
-    auth_url = %{auth_url | query: nil}
-
-    params =
-      params
-      |> map_string_to_atom([:state, :code])
-      |> put_non_nil(:callback_url, to_string(auth_url))
-
-    authenticate(provider, params, opts)
-  end
-
-  @impl true
-  def authenticate(provider, %{code: _code, callback_url: url} = params, opts) do
-    opts = opts ++ @defaults
-    module = Keyword.get(opts, :oauth2_module)
-
-    case validate_options(opts, [:client_id, :client_secret]) do
-      {:ok, _} ->
-        token =
-          params
-          |> Map.take([:code])
-          |> put_non_nil(:redirect_uri, url)
-          |> put_non_nil(:client_id, Keyword.get(opts, :client_id))
-          |> put_non_nil(:client_secret, Keyword.get(opts, :client_secret))
-          |> Enum.into([])
-          |> module.get_token!(opts)
-
-        with {:access_token, token, at} when not is_nil(at) <-
-               {:access_token, token, token.access_token},
-             {:fetch_auth, {:ok, slack_auth}} <- {:fetch_auth, fetch_auth(token)},
-             {:fetch_user, {:ok, slack_user}} <- {:fetch_user, fetch_user(token, slack_auth)},
-             {:fetch_team, {:ok, slack_team}} <- {:fetch_team, fetch_team(token)} do
-          {:ok, construct_auth(provider, token, slack_auth, slack_user, slack_team, opts)}
-        else
-          {:access_token, token, nil} ->
-            {:error,
-             create_failure(provider, __MODULE__, [
-               error(token.other_params["error"], token.other_params["error_description"])
-             ])}
-
-          {_, {:error, %Error{} = err}} ->
-            {:error, create_failure(provider, __MODULE__, [err])}
-        end
-
-      {:error, reason} when is_atom(reason) ->
-        reason_string = to_string(reason)
-        create_failure(provider, __MODULE__, [error(reason, reason_string)])
-    end
-  end
-
-  def authenticate(provider, _, _opts) do
-    {:error,
-     create_failure(provider, __MODULE__, [
-       error(:invalid_callback_params, "invalid callback params")
-     ])}
-  end
-
-  defp construct_auth(provider, token, slack_auth, slack_user, slack_team, opts) do
-    %Auth{
-      provider: provider,
-      strategy: __MODULE__,
-      credentials: credentials(token, slack_auth, slack_user),
-      info: info(slack_auth, slack_user),
-      extra: extra(token, slack_auth, slack_user, slack_team)
-    }
-    |> apply_uid(opts)
-  end
-
-  defp fetch_auth(token) do
-    case Ueberauth.Strategy.Slack.OAuth.get(token, "/auth.test") do
```

Comment:
I refactored all of these nested `case` and `if` statements and rearranged the functions such that public came first and private last, ordered alphabetically.

Reply:
@doomspork will Credo help with this?

Reply:
I do not believe Credo or the formatter know — yet — to look for deeply nested control structures. Credo does flag some of the code

> 8 refactoring opportunities, 71 code readability issues, 10 software design suggestions

Reply:
Got it, I will keep it in mind then, so I can restructure the code before Credo flag is out.

Reply:
Blocks like this we should refactor regardless of Credo or any other formatter:

```elixir
  def authenticate(provider, params, opts) do
    opts = opts ++ @defaults

    with {:ok, _} <- validate_options(opts, [:client_id, :client_secret]) do
      module = Keyword.get(opts, :oauth2_module, @oauth2_module)
      state = Map.get(params, :state)

      with {:token_params, {:ok, token_params}} <-
             {:token_params, fetch_token_params(params, opts)},
           {:token, token} <- module.get_token!(token_params, opts),
           {:access_token, token, at} when not is_nil(at) <-
             {:access_token, token, token.access_token},
           {:state_match, true} <- {:state_match, token_state_matches?(token, state)} do
        token
        |> fetch_user(opts)
        |> build_auth_from_user(token, provider, opts)
      else
        {:token_params, {:error, _}} ->
          {:error,
           create_failure(provider, __MODULE__, [error("missing_code", "No code received")])}

        {:access_token, token, nil} ->
          {:error,
           create_failure(
             provider,
             __MODULE__,
             [error(token.other_params["error"], token.other_params["error_description"])]
           )}

        {:state_match, false} ->
          {:error,
           create_failure(provider, __MODULE__, [error("state_mismatch", "State does not match")])}
      end
    end
  end
```

It's difficult to follow along with what's happening. We create and use some tuples unnecessary. There's a couple other issues within. This type of code exists in each of the strategies. If we can refactor this to be cleaner and clearer with it's intent we may very well identify ways to clean up _all_ of our strategies and simplify them significantly.

Reply:
Making the code legible, but more importantly, approachable by others is very important to me. The easier people find it to get into the Ueberauth (and Guardian) code, the easier it will be to get contributions to fixes and improvements over time.

Reply:
Roger!

Should we try to improve the `guardian_phoenix` before release it? I am not sure what kind of work could be done there but I may be a good idea to go throw and give it some thought.

Any direction on this?

Reply:
@yordis I can take a look later. All that is left is that one PR, correct?

Reply:
@doomspork yes, I will fix the comments tonight and so far in terms of code nothing changed there, just copied over the source code from Guardian.